### PR TITLE
[DEV] Add Native Support for Wayland

### PR DESCRIPTION
**Pull request by AiraYumi**
Enables native support for the Wayland desktop environment.

FLTK 1.4 is still in the prerelease stage, but the API changes are said to be stable, so I would like to start working on it at this stage.

Related Issues : https://github.com/secondlife/viewer/issues/1442 [Linux] Add Native Support for Wayland

- [x] Compile viewer
- [x] Use libdecor for GNOME
- [x] Allows you to launch the viewer.
- [x] Add Wayland Native Dialogs
- [x] Enable IME

### CODE REVIEW

**Comment by AiraYumi** [comment]
@Nicky-D 
If I were to start working on it, would it be better to merge the "feature/wayland_compat" branch?
Or is it too fast to start the work now?

**Comment by RyeMutt** [comment]
I've had good success with using https://github.com/btzy/nativefiledialog-extended for wayland-native dialogs under Alchemy via gtk3 or dbus portals.

**Comment by AiraYumi** [comment]
Thank you @Nicky-D and @RyeMutt .

I don't think FLTK 1.4 will be released anytime soon, so I'd like to take my time to resolve the issue.

**Comment by Nicky-D** [comment]
https://github.com/Nicky-D/viewer/tree/feature/wayland_compat is no up to date (with the caveats mentioned above)

**Comment by AiraYumi** [comment]
Thank you. On the fltk side, I referred to the fltk documentation and included wayland-related libraries as dependencies.

File: indra/cmake/UI.cmake [hunk 1]
```diff
@@ -25,6 +25,8 @@ if (LINUX)
 
   target_link_libraries( ll::uilibraries INTERFACE
           fltk
+          pangocairo
+          wayland-client
```
Comment (Nicky-D) [inline indra/cmake/UI.cmake:29]: Wayland client is handled via pkgconfig (line 17). You can comment it in in line 20. It's commented out to find a better way, but for this draft there is no reason to not enable it

Comment (AiraYumi) [reply indra/cmake/UI.cmake]: Thank you.
The changes have been reflected.

**Comment by bennettgoble** [comment]
@AiraYumi thanks for the work so far on this. I have some feedback on the solution here (And LL could be willing to help.) Could you make an issue outlining the requested enhancement so that the solution can be discussed before implementation?

**Comment by bennettgoble** [comment]
> I've had good success with using https://github.com/btzy/nativefiledialog-extended for wayland-native dialogs under Alchemy via gtk3 or dbus portals. I was considering porting it over since portals eliminate most dependency on native distro libraries.

If this route reduces system dependencies and increases system compatibility it could be a good solution?

**Comment by AiraYumi** [comment]
@bennettgoble The relevant issue can be found here.

[Linux] Add Native Support for Wayland
https://github.com/secondlife/viewer/issues/1442

**Comment by AiraYumi** [comment]
oops. 
Something broke when I merged the develop branch.
Can someone check this?

```
/home/runner/work/viewer/viewer/indra/llwindow/llwindowsdl.cpp:1224:9: error: use of undeclared identifier 'mReallyCapturedCount'
    if (mReallyCapturedCount <= 0) // uncapture
        ^
/home/runner/work/viewer/viewer/indra/llwindow/llwindowsdl.cpp:1232:9: error: use of undeclared identifier 'mReallyCapturedCount'
    if (mReallyCapturedCount < 0) // yuck, imbalance.
        ^
/home/runner/work/viewer/viewer/indra/llwindow/llwindowsdl.cpp:1234:9: error: use of undeclared identifier 'mReallyCapturedCount'
        mReallyCapturedCount = 0;
        ^
3 errors generated.
```

**Comment by AiraYumi** [comment]
Tagging the source code that was successfully compiled.
https://github.com/AiraYumi/viewer/releases/tag/Compiled_Linux_Enable_Wayland

**Comment by AiraYumi** [comment]
I was able to compile and run it, but I don't know if it's running on wayland or xwayland.

**Comment by Nicky-D** [comment]
> I was able to compile and run it, but I don't know if it's running on wayland or xwayland.

You will find a line "Running under Wayland" in the logs.
In a wayland session you can trigger this be either
`SDL_VIDEODRIVER="wayland" ./secondlife`

or
`DISPLAY= ./secondlife`

**Comment by AiraYumi** [comment]
The viewer can now be built. I think that most of what can be done at this point is done.
All that remains is to add a dependency on libdecor, adjust library dependencies, and fix potential issues.
Once a release GHA build is available, we will be able to do more thorough testing.

**Comment by AiraYumi** [comment]
@bennettgoble @marchcat If possible, could you please do the build here too?

**Comment by Nicky-D** [comment]
@AiraYumi the pr is slightly outdated 
- there's a lot of stuff in there that had no business to be here (I accidentally committed clion settings and my cmakepresets). I delete those files in the base branch 
- my changes now dynamically load libewayland-client.so. thus dropping the hard dependency. This is important for pure x11 installations

**Comment by AiraYumi** [comment]
Thank you. @Nicky-D 

Once you've finished making changes, please let us know and we'll merge them here.
If necessary, please send a pull request to my repository and I will check and merge them.
There are probably unnecessary dependencies or functions in fltk and sdl2, so I would appreciate it if you could check them.

https://github.com/AiraYumi/viewer/tree/linux/enable_wayland
https://github.com/AiraYumi/3p-open-libndofdev/tree/with-libdecor
https://github.com/AiraYumi/3p-sdl2/tree/update-with-libdecor
https://github.com/AiraYumi/3p-fltk/tree/develop

**Comment by AiraYumi** [comment]
To build fltk that is compatible with wayland, you need to depend on libwayland-dev, so viewers that use fltk also need to depend on libwayland-client.
I wish there was another way...

~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /root/viewer/build-linux-x86_64/packages/lib/release/libfltk.a(Fl_Wayland_Screen_Driver.o): undefined reference to symbol 'wl_output_interface'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib64/libwayland-client.so.0: error adding symbols: DSO missing from command line
~~~

**Comment by Nicky-D** [comment]
Afaik Rye wants to replace FTLK with NFDE in the near future. This will break the FTLK dependency on libwaylandclient.

As this is all very much WIP it is best to keep an eye out on my repository and this one. For example you really want #2773
as well. As the other SDL changes in develop broke Wayland support (or maybe the combination Wayland/AMD) and this PR fixes that problem.

**Comment by AiraYumi** [comment]
oops.
Something broke in the latest merge and I need to find out why.

**Comment by AiraYumi** [comment]
fix it.

**Comment by AiraYumi** [comment]
It doesn't seem to work on wayland. It crashes. Now.

**Comment by Nicky-D** [comment]
Try to merge develop again. That shall fix it

**Comment by AiraYumi** [comment]
I merged it, compiled it locally and it worked.

**Comment by AiraYumi** [comment]
I think it probably works on ReleaseOS. But I can't confirm if it works fine on Release build.
Any ideas?

**Comment by bennettgoble** [comment]
I don't want all this work to be done and it to just get tossed out because there is an alternative route with less dependencies. On its face, bringing in FLTK for file dialogs seems heavy handed. @AiraYumi would you be interested in pursuing adding Wayland support using the methods described by @RyeMutt? @RyeMutt would you be willing to provide some more information about what work would be required? (And add the details to #1442?)

**Comment by bennettgoble** [comment]
Also: I love you all. We really ought to figure out how to have a real-time venue for OSS collaboration.

**Comment by Nicky-D** [comment]
@bennettgoble Linux uses FLTK right now, it's not about bringing it it, but tossing it out and using something else with less dependencies.

For the sake of granularity I created #2803 this lays the foundating AiraYumi already build on and removed the problem of them not being allowed to contribute my code.

I suggest a second PR for the updated SDL2, afaik this also add no new dependency and adds the window decorations via SDL, so we do not have to mess with libdecor (which is gnome only anyway) ourselves.

The third and final step would be FLTK vs NFDE. The viewer runs without it. But there will be no (pure Wayland) file or colorpickers without one of the two

**Comment by AiraYumi** [comment]
@bennettgoble I'm interested.

To be honest, I'm struggling with how to resolve this RP. As far as I can tell, it seems like there's no way to avoid the need for libwayland-client to build the viewer when using FLTK.
This is because fltk requires libwayland-client to enable wayland.
So no matter how much you enable dynamic loading on the viewer side, the situation remains that libwayland-client is required when building.

If FLTK is only being used to create file dialogs, it seems unnatural to continue using it solely for that purpose.

**Comment by AiraYumi** [comment]
broken. after merge...

**Comment by AiraYumi** [comment]
Hold off on merging the develop branch until you are ready to do a linux build on the develop branch.

**Comment by AiraYumi** [comment]
We have reverted FLTK to 1.3.9 to avoid adding unnecessary build-time and run-time dependencies.
Additionally, I've incorporated wayland_base_support provided by @Nicky-D.

**Comment by AiraYumi** [comment]
Currently it fails to build with clang(In my environment).
It builds successfully with gcc.
Also, it occasionally freezes when reading the world.

**Comment by AiraYumi** [comment]
Linux build cannot be generated because there is no configuration setting. We added the setting for this PR only.

**Comment by AiraYumi** [comment]
Oops.linux build broken...

**Comment by AiraYumi** [comment]
@RyeMutt @bennettgoble @marchcat, I would like to incorporate NFDE into this PR, so could you please create a 3p package?

Or do you have any other good ideas?

**Comment by marchcat** [comment]
@AiraYumi, https://github.com/btzy/nativefiledialog-extended this one?

**Comment by marchcat** [comment]
I haven't researched this much yet, so just asking about the plan.

**Comment by Quinn-Elara** [comment]
I've just tested the latest build from this pull on my Fedora 40 machine, however the viewer fails to launch with the following errors:
`2024-10-11T13:26:44Z ERROR # llwindow/llwindowsdl.cpp(518) createContext : Failed to load wayland-client.so or grab required functions
2024-10-11T13:26:44Z WARNING # llwindow/llwindow.cpp(75) OSMessageBox : OSMessageBox: 2024-10-11T13:26:44Z ERROR # llwindow/llwindowsdl.cpp(518) createContext : Failed to load wayland-client.so or grab required functions


This is a developer-only notification!
This notification won't be present in Release for download build
zenity: symbol lookup error: /lib64/libgtk-4.so.1: undefined symbol: vkCreateXlibSurfaceKHR
2024-10-11T13:26:44Z INFO # llcommon/llapp.cpp(601) default_unix_signal_handler : Signal handler - Got signal 11 - Segmentation fault
2024-10-11T13:26:44Z WARNING # llcommon/llapp.cpp(674) default_unix_signal_handler : Signal handler - Handling fatal signal!
2024-10-11T13:26:44Z WARNING # llcommon/llapp.cpp(693) default_unix_signal_handler : Signal handler - Flagging error status and waiting for shutdown
2024-10-11T13:26:44Z INFO # llcommon/llapp.cpp(374) setStatus : status: error
2024-10-11T13:26:44Z WARNING # llcommon/llapp.cpp(709) default_unix_signal_handler : Signal handler - App is stopped, reraising signal
./secondlife: line 87: 187110 Segmentation fault      (core dumped) $LL_WRAPPER bin/do-not-directly-run-secondlife-bin "${ARGS[@]}"
*** Bad shutdown ($LL_RUN_ERR). ***
`
The viewer **will** run if executed via `sudo`, however this isn't really safe for general use.
Seems like it's not able to load `libwayland-client` when run with default user permissions. (The package is installed!)

**Comment by AiraYumi** [comment]
Are you self-compiling?
Did you use a pre-built version?

If it is self compiling, we do not know why it does not work.
If you used a pre-built version, you can probably fix it.

**Comment by Quinn-Elara** [comment]
> Are you self-compiling? Did you use a pre-built version?
> 
> If it is self compiling, we do not know why it does not work. If you used a pre-built version, you can probably fix it.

This is using the prebuilt version from this GHA run: https://github.com/secondlife/viewer/actions/runs/11286008602

**Comment by AiraYumi** [comment]
I think it's probably cured. 

this GHA
https://github.com/secondlife/viewer/actions/runs/11303216218

**Comment by Quinn-Elara** [comment]
> I think it's probably cured.
> 
> this GHA https://github.com/secondlife/viewer/actions/runs/11314217447

Just had a try with this build, same error unfortunately:
```
2024-10-14T11:24:25Z ERROR # llwindow/llwindowsdl.cpp(518) createContext : Failed to load wayland-client.so or grab required functions
2024-10-14T11:24:25Z WARNING # llwindow/llwindow.cpp(75) OSMessageBox : OSMessageBox: 2024-10-14T11:24:25Z ERROR # llwindow/llwindowsdl.cpp(518) createContext : Failed to load wayland-client.so or grab required functions


This is a developer-only notification!
This notification won't be present in Release for download build
zenity: symbol lookup error: /lib64/libgtk-4.so.1: undefined symbol: vkCreateXlibSurfaceKHR
2024-10-14T11:24:25Z INFO # llcommon/llapp.cpp(601) default_unix_signal_handler : Signal handler - Got signal 11 - Segmentation fault
2024-10-14T11:24:25Z WARNING # llcommon/llapp.cpp(674) default_unix_signal_handler : Signal handler - Handling fatal signal!
2024-10-14T11:24:25Z WARNING # llcommon/llapp.cpp(693) default_unix_signal_handler : Signal handler - Flagging error status and waiting for shutdown
2024-10-14T11:24:25Z INFO # llcommon/llapp.cpp(374) setStatus : status: error
2024-10-14T11:24:25Z WARNING # llcommon/llapp.cpp(709) default_unix_signal_handler : Signal handler - App is stopped, reraising signal
./secondlife: line 87: 24327 Segmentation fault      (core dumped) $LL_WRAPPER bin/do-not-directly-run-secondlife-bin "${ARGS[@]}"
*** Bad shutdown ($LL_RUN_ERR). ***
```

**Comment by AiraYumi** [comment]
Under investigation

**Comment by AiraYumi** [comment]
Changed to read "libwayland-client.so.0" instead of "libwayland-client.so".
Also changed SDL to use x11 for drawing when wayland-client cannot be read.

**Comment by AiraYumi** [comment]
As the commit history has grown, I plan to create a new pull request once I have confirmed that it can be run on KDE and GNOME.

**Comment by Quinn-Elara** [comment]
Can confirm that the latest build (https://github.com/secondlife/viewer/actions/runs/11339026891) does run on my system without errors, however it's running under X11 with (seemingly) no attempt to run under Wayland by default.

Excerpt of console log (it does continue after this, but this is from the same stage as where it was previously erroring out)
```
2024-10-15T10:55:13Z INFO #Window# llwindow/llwindowsdl.cpp(591) toggleVSync : Enabling vertical sync
2024-10-15T10:55:13Z INFO # llwindow/llwindowsdl.cpp(513) createContext : Running under X11
2024-10-15T10:55:13Z INFO # llrender/llgl.cpp(2596) parse_gl_version : GL: 4.6 (Compatibility Profile) Mesa 24.1.7
```

If I run the shell script with `SDL_VIDEODRIVER=wayland ./secondlife`, the viewer launches under Wayland without issue (without the `wayland-devel` package installed!). Log excerpt:
```
2024-10-15T11:00:00Z INFO #Window# llwindow/llwindowsdl.cpp(591) toggleVSync : Enabling vertical sync
2024-10-15T11:00:00Z INFO # llwindow/llwindowsdl.cpp(533) createContext : Running under Wayland
2024-10-15T11:00:00Z WARNING # llwindow/llwindowsdl.cpp(534) createContext : Be aware that with at least SDL2 the window will not receive minimizing events, thus minimized state can only be estimated.also setting the application icon via SDL_SetWindowIcon does not work.
2024-10-15T11:00:00Z INFO # llrender/llgl.cpp(2596) parse_gl_version : GL: 4.6 (Compatibility Profile) Mesa 24.1.7
```

**Comment by AiraYumi** [comment]
With this change it should probably work without having to set any environment variables.

**Comment by Quinn-Elara** [comment]
> With this change it should probably work without having to set any environment variables.

Can confirm that the latest build (https://github.com/secondlife/viewer/actions/runs/11345628939) runs on Wayland out of the box without issue on my system.
```
2024-10-15T13:00:06Z INFO #Window# llwindow/llwindowsdl.cpp(591) toggleVSync : Enabling vertical sync
2024-10-15T13:00:06Z INFO # llwindow/llwindowsdl.cpp(533) createContext : Running under Wayland
2024-10-15T13:00:06Z WARNING # llwindow/llwindowsdl.cpp(534) createContext : Be aware that with at least SDL2 the window will not receive minimizing events, thus minimized state can only be estimated.also setting the application icon via SDL_SetWindowIcon does not work.
2024-10-15T13:00:06Z INFO # llrender/llgl.cpp(2596) parse_gl_version : GL: 4.6 (Compatibility Profile) Mesa 24.1.7
```

**Comment by AiraYumi** [comment]
I'm thinking of starting work on incorporating NFDE into the viewer. What do you think?

**Comment by AiraYumi** [comment]
I'm not sure if libdecor is working. Can I do it just by modifying SDL2?

**Comment by AiraYumi** [comment]
A new Pull Request #2927 has been created.

This will be closed once #2927 is merged.

**Comment by Quinn-Elara** [comment]
I've noticed when the viewer is running under Wayland, changing any settings that cause the viewer to recompile shaders (e.g. enabling / disabling SSR or Mirrors) will cause the viewer to stop drawing new frames (freezing on the last frame). The viewer hasn't crashed, and the UI can still be navigated if you can remember where the buttons are / use key combos. This persists until the viewer is restarted, at which point it (usually) runs normally.

**Comment by AiraYumi** [comment]
> I've noticed when the viewer is running under Wayland, changing any settings that cause the viewer to recompile shaders (e.g. enabling / disabling SSR or Mirrors) will cause the viewer to stop drawing new frames (freezing on the last frame). The viewer hasn't crashed, and the UI can still be navigated if you can remember where the buttons are / use key combos. This persists until the viewer is restarted, at which point it (usually) runs normally.

I don't know what the cause is.
I've incorporated the sdl2 fix.

**Comment by AiraYumi** [comment]
There are still some bugs remaining, but we believe that all the work that can be done at this point is complete.

**Comment by AiraYumi** [comment]
This PR is closed as the planned work is complete.

**Pull request closed without merging**